### DISCUSSION TITLE
I got RESOURCE_EXHAUSTED but I still have memory

### DISCUSSION THREAD

**Discussion by kkeevin123456** [jax-ml/jax#10008] [Q&A]
Here is the question,
I have 4 gpus but jax always allocate kernel matrix to single device.
How could I use all gpus?

![image](https://user-images.githubusercontent.com/55287979/159721012-71d4135a-e17b-4f89-91cb-e2cd18e30a5c.png)
![image](https://user-images.githubusercontent.com/55287979/159721061-6457b62d-2879-42b8-aad7-36886baa1e44.png)

Thanks for help.

**Comment by YouJiacheng** [comment]
You need `pmap` to leverage multiple devices. 
https://jax.readthedocs.io/en/latest/_autosummary/jax.pmap.html
https://jax.readthedocs.io/en/latest/jax-101/06-parallelism.html
FYI, we cannot use `vmap` for multiple devices because of the non-unified memory nature of multiple devices, thus inter device parallel is different from intra device parallel.
It just like in pytorch, we need to use DDP or DP for multiple devices.

**Comment by kkeevin123456** [comment]
Thanks for reply.
I want to apply pmap on empirical_ntk_fn like below.

```
eNTK = nt.empirical_ntk_fn(apply_fn, vmap_axes=0, trace_axes=(), implementation=2)
gram = eNTK(inputs, None, get_params(opt_state))
```

How can I pass the params?
My idea is like this, but it still has some bug.

```
eNTK = nt.empirical_ntk_fn(apply_fn, vmap_axes=0, trace_axes=(), implementation=2)
gram = pmap(eNTK)(inputs, None, get_params(opt_state))
```

![image](https://user-images.githubusercontent.com/55287979/159728309-ff06f64b-901f-4466-b4b7-1dc5420606c7.png)

**Reply by YouJiacheng** [reply]
Could you give a runnable code snippet?
I have no idea what `nt.empirical_ntk_fn` is.

**Comment by kkeevin123456** [comment]
Here is the code that can reproduce the problem~
The code below is to load data and train a toy CNN.
Here I want to calculate the kernel matrix of first layer that will cause the problem.

```
import numpy as np
import torch
import torchvision.datasets as datasets
import torch.nn.functional as F

import jax
from jax import random
import jax.numpy as jnp
from jax.example_libraries import optimizers
from jax import jit, grad, vmap, pmap
import functools
import neural_tangents as nt
from neural_tangents import stax

from tqdm import tqdm
```

```
mnist_trainset = datasets.MNIST(root='./data', train=True, download=True, transform=None)
class_5_idx = np.where(mnist_trainset.targets.numpy() == 5)
class_3_idx = np.where(mnist_trainset.targets.numpy() == 3)
class_5 = mnist_trainset.data[class_5_idx]
class_3 = mnist_trainset.data[class_3_idx]
```

```
# reproduce problem
M = 200
W = H = 28
C = 1
P = 200
eta = 0.1

inputs = np.vstack((class_5[:P//2], class_3[:P//2]))
idx = np.arange(P)
np.random.shuffle(idx)
inputs = inputs[idx].reshape(-1, 28, 28, 1).astype('float32')
inputs = jnp.array(inputs)

ys = np.vstack((*np.ones(P//2), *np.zeros(P//2)))
ys = ys[idx].astype('float32')
ys = jnp.array(ys)

# kernel for all layers
init_fn, apply_fn, kernel_fn = stax.serial(
    stax.Conv(10, (5, 5), (1, 1), 'SAME'), stax.Relu(),  # shape (output_c, filter_size, stride_size)
    stax.Flatten(), stax.Dense(1)
)

# kernel for first layers
init_fn_a, apply_fn_a, kernel_fn_a = stax.serial(
    stax.Conv(10, (5, 5), (1, 1), 'SAME'), stax.Relu(),  # shape (output_c, filter_size, stride_size)
    stax.Flatten()
)

shape, params = init_fn(random.PRNGKey(np.random.randint(1e6)), inputs.shape)
eNTK = nt.empirical_ntk_fn(apply_fn, vmap_axes=0, trace_axes=(), implementation=2) # implementation 2 會比較快，1會比較慢但比較精確

opt_init, opt_update, get_params = optimizers.sgd(eta)
opt_state = opt_init(params)

eNTK_a = nt.empirical_ntk_fn(apply_fn_a, vmap_axes=0, trace_axes=(), implementation=2)

loss = jit(lambda params, x, y: 0.5 * jnp.mean((apply_fn(params, x) - y) ** 2))
grad_loss = jit(lambda state, x, y: grad(loss)(get_params(state), x, y))

nsteps = 1000

for i in range(nsteps):
    opt_state = opt_update(i, grad_loss(opt_state, inputs, ys), opt_state)
    gram_a = eNTK_a(inputs, None, [get_params(opt_state)[0], get_params(opt_state)[1], get_params(opt_state)[2]])
    print(gram_a.shape)
```

**Reply by YouJiacheng** [reply]
Ok, you should pmap over inputs only.

**Reply by kkeevin123456** [reply]
When I try to pmap over inputs only it cause this error.

![image](https://user-images.githubusercontent.com/55287979/160040186-724d189f-72bb-402b-8885-9291ff7b1d96.png)

Then I try to pmap another term which will alse cause error.

![image](https://user-images.githubusercontent.com/55287979/160040236-aea568d6-e12b-4ca4-aae6-c41ddf7611a4.png)
![image](https://user-images.githubusercontent.com/55287979/160040278-cff258aa-74c0-4804-a4e3-075eac64bfbb.png)

**Reply by YouJiacheng** [reply]
Emmm.
```python
gram_a = pmap(lambda ins: eNTK_a(ins, None, [get_params(opt_state)[0], get_params(opt_state)[1], get_params(opt_state)[2]]))(inputs)
```

**Reply by kkeevin123456** [reply]
I use your code but it cause this error.
![image](https://user-images.githubusercontent.com/55287979/160086507-1139fa4e-e257-433c-b95d-4c2e2ce5f81b.png)

So I change into this:

```
gram_a = pmap(lambda ins: eNTK_a(ins, None, [get_params(opt_state)[0], get_params(opt_state)[1], get_params(opt_state)[2]]))(np.array([inputs]))
```

But it still put all data into one device.
![image](https://user-images.githubusercontent.com/55287979/160086723-357ef583-eb74-4dd7-97ec-e83476a7918e.png)

I'm sure the jax can see all devices.
![image](https://user-images.githubusercontent.com/55287979/160086829-dec0375e-68c8-4e53-9d3c-7e55940e05dd.png)

**Reply by YouJiacheng** [reply]
What is the shape of inputs.
